File: oai-cn5g-nwdaf/components/oai-nwdaf-sbi/internal/sbi/nrf_client.go
/*
 * Licensed to the OpenAirInterface (OAI) Software Alliance under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The OpenAirInterface Software Alliance licenses this file to You under
 * the OAI Public License, Version 1.1  (the "License"); you may not use this
 * file except in compliance with the License. You may obtain a copy of the
 * License at
 *
 *      http://www.openairinterface.org/?page_id=698
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *-------------------------------------------------------------------------------
 * For more information about the OpenAirInterface (OAI) Software Alliance:
 *      contact@openairinterface.org
 */

/*
 * Description: NRF client for NWDAF registration (3GPP TS 29.510)
 */

package sbi

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"time"
)

// NFProfile represents the NF Profile for NRF registration
type NFProfile struct {
	NFInstanceID    string       `json:"nfInstanceId"`
	NFType          string       `json:"nfType"`
	NFStatus        string       `json:"nfStatus"`
	HeartBeatTimer  int          `json:"heartBeatTimer"`
	IPv4Addresses   []string     `json:"ipv4Addresses,omitempty"`
	NFServices      []NFService  `json:"nfServices,omitempty"`
	NWDAFInfo       *NWDAFInfo   `json:"nwdafInfo,omitempty"`
}

// NFService represents an NF service
type NFService struct {
	ServiceInstanceID string       `json:"serviceInstanceId"`
	ServiceName       string       `json:"serviceName"`
	Versions          []NFVersion  `json:"versions"`
	Scheme            string       `json:"scheme"`
	NFServiceStatus   string       `json:"nfServiceStatus"`
	IPEndPoints       []IPEndPoint `json:"ipEndPoints,omitempty"`
}

// NFVersion represents service version
type NFVersion struct {
	APIVersionInURI string `json:"apiVersionInUri"`
	APIFullVersion  string `json:"apiFullVersion"`
}

// IPEndPoint represents IP endpoint
type IPEndPoint struct {
	IPv4Address string `json:"ipv4Address,omitempty"`
	Port        int    `json:"port,omitempty"`
}

// NWDAFInfo contains NWDAF-specific information
type NWDAFInfo struct {
	EventIDs    []string `json:"eventIds,omitempty"`
	NWDAFEvents []string `json:"nwdafEvents,omitempty"`
}

// NRFClient handles NRF registration
type NRFClient struct {
	nrfEndpoint      string
	nfInstanceID     string
	ipv4Address      string
	sbiPort          int
	heartbeatTimer   int
	autoRegister     bool
	registered       bool
	stopHeartbeat    chan struct{}
}

// NewNRFClient creates a new NRF client
func NewNRFClient(nrfEndpoint, nfInstanceID, ipv4Address string, sbiPort, heartbeatTimer int, autoRegister bool) *NRFClient {
	return &NRFClient{
		nrfEndpoint:    nrfEndpoint,
		nfInstanceID:   nfInstanceID,
		ipv4Address:    ipv4Address,
		sbiPort:        sbiPort,
		heartbeatTimer: heartbeatTimer,
		autoRegister:   autoRegister,
		stopHeartbeat:  make(chan struct{}),
	}
}

// buildNFProfile creates the NF Profile for registration
func (c *NRFClient) buildNFProfile() NFProfile {
	return NFProfile{
		NFInstanceID:   c.nfInstanceID,
		NFType:         "NWDAF",
		NFStatus:       "REGISTERED",
		HeartBeatTimer: c.heartbeatTimer,
		IPv4Addresses:  []string{c.ipv4Address},
		NFServices: []NFService{
			{
				ServiceInstanceID: fmt.Sprintf("%s-nnwdaf-eventssubscription", c.nfInstanceID),
				ServiceName:       "nnwdaf-eventssubscription",
				Versions: []NFVersion{
					{
						APIVersionInURI: "v1",
						APIFullVersion:  "1.0.0",
					},
				},
				Scheme:          "http",
				NFServiceStatus: "REGISTERED",
				IPEndPoints: []IPEndPoint{
					{
						IPv4Address: c.ipv4Address,
						Port:        c.sbiPort,
					},
				},
			},
			{
				ServiceInstanceID: fmt.Sprintf("%s-nnwdaf-analyticsinfo", c.nfInstanceID),
				ServiceName:       "nnwdaf-analyticsinfo",
				Versions: []NFVersion{
					{
						APIVersionInURI: "v1",
						APIFullVersion:  "1.0.0",
					},
				},
				Scheme:          "http",
				NFServiceStatus: "REGISTERED",
				IPEndPoints: []IPEndPoint{
					{
						IPv4Address: c.ipv4Address,
						Port:        c.sbiPort,
					},
				},
			},
		},
		NWDAFInfo: &NWDAFInfo{
			EventIDs: []string{
				"NF_LOAD",
				"UE_MOBILITY",
				"UE_COMM",
				"QOS_SUSTAINABILITY",
				"ABNORMAL_BEHAVIOUR",
			},
			NWDAFEvents: []string{
				"NF_LOAD",
				"UE_MOBILITY",
			},
		},
	}
}

// Register registers the NWDAF with NRF
func (c *NRFClient) Register() error {
	if !c.autoRegister {
		log.Println("NRF auto-registration disabled")
		return nil
	}

	if c.nrfEndpoint == "" {
		log.Println("NRF endpoint not configured, skipping registration")
		return nil
	}

	nfProfile := c.buildNFProfile()
	profileJSON, err := json.Marshal(nfProfile)
	if err != nil {
		return fmt.Errorf("failed to marshal NF profile: %v", err)
	}

	url := fmt.Sprintf("%s/nnrf-nfm/v1/nf-instances/%s", c.nrfEndpoint, c.nfInstanceID)
	log.Printf("Registering NWDAF with NRF at %s", url)

	client := &http.Client{Timeout: 10 * time.Second}
	req, err := http.NewRequest(http.MethodPut, url, bytes.NewBuffer(profileJSON))
	if err != nil {
		return fmt.Errorf("failed to create request: %v", err)
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := client.Do(req)
	if err != nil {
		return fmt.Errorf("failed to register with NRF: %v", err)
	}
	defer resp.Body.Close()

	body, _ := ioutil.ReadAll(resp.Body)

	if resp.StatusCode == http.StatusOK || resp.StatusCode == http.StatusCreated {
		c.registered = true
		log.Printf("Successfully registered with NRF (status: %d)", resp.StatusCode)

		// Start heartbeat
		go c.heartbeatLoop()
		return nil
	}

	return fmt.Errorf("NRF registration failed with status %d: %s", resp.StatusCode, string(body))
}

// Deregister removes the NWDAF from NRF
func (c *NRFClient) Deregister() error {
	if !c.registered {
		return nil
	}

	// Stop heartbeat
	close(c.stopHeartbeat)

	url := fmt.Sprintf("%s/nnrf-nfm/v1/nf-instances/%s", c.nrfEndpoint, c.nfInstanceID)

	client := &http.Client{Timeout: 10 * time.Second}
	req, err := http.NewRequest(http.MethodDelete, url, nil)
	if err != nil {
		return fmt.Errorf("failed to create deregister request: %v", err)
	}

	resp, err := client.Do(req)
	if err != nil {
		return fmt.Errorf("failed to deregister from NRF: %v", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusOK || resp.StatusCode == http.StatusNoContent {
		c.registered = false
		log.Println("Successfully deregistered from NRF")
		return nil
	}

	return fmt.Errorf("NRF deregistration failed with status %d", resp.StatusCode)
}

// heartbeatLoop sends periodic heartbeats to NRF
func (c *NRFClient) heartbeatLoop() {
	ticker := time.NewTicker(time.Duration(c.heartbeatTimer) * time.Second)
	defer ticker.Stop()

	for {
		select {
		case <-c.stopHeartbeat:
			return
		case <-ticker.C:
			c.sendHeartbeat()
		}
	}
}

// sendHeartbeat sends a heartbeat to NRF
func (c *NRFClient) sendHeartbeat() {
	url := fmt.Sprintf("%s/nnrf-nfm/v1/nf-instances/%s", c.nrfEndpoint, c.nfInstanceID)

	// PATCH with heartbeat update
	patchData := []map[string]interface{}{
		{
			"op":    "replace",
			"path":  "/nfStatus",
			"value": "REGISTERED",
		},
	}

	patchJSON, err := json.Marshal(patchData)
	if err != nil {
		log.Printf("Failed to marshal heartbeat patch: %v", err)
		return
	}

	client := &http.Client{Timeout: 10 * time.Second}
	req, err := http.NewRequest(http.MethodPatch, url, bytes.NewBuffer(patchJSON))
	if err != nil {
		log.Printf("Failed to create heartbeat request: %v", err)
		return
	}
	req.Header.Set("Content-Type", "application/json-patch+json")

	resp, err := client.Do(req)
	if err != nil {
		log.Printf("NRF heartbeat failed: %v", err)
		return
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusOK {
		log.Println("NRF heartbeat successful")
	} else if resp.StatusCode == http.StatusNotFound {
		log.Println("NF not found in NRF, re-registering...")
		c.registered = false
		c.Register()
	} else {
		log.Printf("NRF heartbeat failed with status %d", resp.StatusCode)
	}
}

// IsRegistered returns the registration status
func (c *NRFClient) IsRegistered() bool {
	return c.registered
}

// Global NRF client instance
var nrfClient *NRFClient

// InitNRFRegistration initializes and performs NRF registration
func InitNRFRegistration() {
	if config.Nrf.Endpoint == "" {
		log.Println("NRF endpoint not configured, skipping NRF registration")
		return
	}

	nrfClient = NewNRFClient(
		config.Nrf.Endpoint,
		config.Nrf.NFInstanceID,
		config.Nrf.IPv4Address,
		config.Nrf.SBIPort,
		config.Nrf.HeartbeatTimer,
		config.Nrf.AutoRegister,
	)

	if err := nrfClient.Register(); err != nil {
		log.Printf("Failed to register with NRF: %v", err)
	}
}

// NFInstancesResponse represents the response from NRF discovery
type NFInstancesResponse struct {
	NFInstances []NFProfile `json:"nfInstances"`
}

// DiscoverNFInstances queries NRF to discover NF instances of a given type
func DiscoverNFInstances(nrfEndpoint string, nfType string, serviceName string) ([]NFProfile, error) {
	if nrfEndpoint == "" {
		return nil, fmt.Errorf("NRF endpoint not configured")
	}

	// Build discovery URL with query parameters
	// GET /nnrf-disc/v1/nf-instances?target-nf-type=NWDAF&service-names=ndccf-datamanagement
	url := fmt.Sprintf("%s/nnrf-disc/v1/nf-instances?target-nf-type=%s", nrfEndpoint, nfType)
	if serviceName != "" {
		url = fmt.Sprintf("%s&service-names=%s", url, serviceName)
	}

	log.Printf("Discovering NF instances from NRF: %s", url)

	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Get(url)
	if err != nil {
		return nil, fmt.Errorf("failed to query NRF: %v", err)
	}
	defer resp.Body.Close()

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response body: %v", err)
	}

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("NRF discovery failed with status %d: %s", resp.StatusCode, string(body))
	}

	var response NFInstancesResponse
	if err := json.Unmarshal(body, &response); err != nil {
		return nil, fmt.Errorf("failed to unmarshal NRF response: %v", err)
	}

	log.Printf("Discovered %d NF instances", len(response.NFInstances))
	return response.NFInstances, nil
}

// DiscoverDCCF discovers DCCF service endpoint via NRF
// It looks for NWDAF instances that provide ndccf-datamanagement service
func DiscoverDCCF(nrfEndpoint string, ownInstanceID string) (string, error) {
	if nrfEndpoint == "" {
		return "", fmt.Errorf("NRF endpoint not configured")
	}

	// Discover NWDAF instances with ndccf-datamanagement service
	instances, err := DiscoverNFInstances(nrfEndpoint, "NWDAF", "ndccf-datamanagement")
	if err != nil {
		return "", fmt.Errorf("failed to discover DCCF: %v", err)
	}

	// Find a DCCF instance (not ourselves)
	for _, instance := range instances {
		// Skip our own instance
		if instance.NFInstanceID == ownInstanceID {
			continue
		}

		// Look for ndccf-datamanagement service
		for _, service := range instance.NFServices {
			if service.ServiceName == "ndccf-datamanagement" && service.NFServiceStatus == "REGISTERED" {
				// Build endpoint URL from service info
				if len(service.IPEndPoints) > 0 {
					ep := service.IPEndPoints[0]
					endpoint := fmt.Sprintf("%s://%s:%d", service.Scheme, ep.IPv4Address, ep.Port)
					log.Printf("Discovered DCCF at %s (instance: %s)", endpoint, instance.NFInstanceID)
					return endpoint, nil
				}
				// Fallback to NFProfile IPv4 addresses
				if len(instance.IPv4Addresses) > 0 {
					endpoint := fmt.Sprintf("http://%s:8081", instance.IPv4Addresses[0])
					log.Printf("Discovered DCCF at %s (instance: %s)", endpoint, instance.NFInstanceID)
					return endpoint, nil
				}
			}
		}
	}

	return "", fmt.Errorf("no DCCF instance found in NRF")
}

// GetNRFEndpoint returns the configured NRF endpoint
func GetNRFEndpoint() string {
	return config.Nrf.Endpoint
}

// GetOwnInstanceID returns the own NF instance ID
func GetOwnInstanceID() string {
	return config.Nrf.NFInstanceID
}

File: oai-cn5g-nwdaf/components/oai-nwdaf-dccf/src/dccf/nrf_client.py
#!/usr/bin/env python3
"""
NRF Client for NWDAF-DCCF NF Registration
Implements 3GPP TS 29.510 NFManagement API
"""
import logging
import asyncio
import os
import json
import subprocess
from typing import Optional

logger = logging.getLogger(__name__)

class NRFClient:
    """NRF Client for NF Registration and Heartbeat"""

    def __init__(self, config: dict):
        self.nrf_endpoint = config.get('nrf', {}).get('endpoint', '')
        self.auto_register = config.get('nrf', {}).get('auto_register', False)
        self.heartbeat_interval = config.get('nrf', {}).get('heartbeat_interval', 10)

        # NWDAF NF Profile
        nwdaf_config = config.get('nwdaf', {})
        self.nf_instance_id = nwdaf_config.get('nfInstanceId', '58f8ec8c-8a94-5ab1-86f6-7c989cb0cb74')
        self.nf_type = nwdaf_config.get('nfType', 'NWDAF')
        self.ipv4_address = nwdaf_config.get('ipv4Address') or os.getenv('POD_IP', '')
        self.nf_services = nwdaf_config.get('nfServices', [])

        self._registered = False
        self._heartbeat_task: Optional[asyncio.Task] = None

    def _curl_json(self, method: str, url: str, payload=None, content_type: str = "application/json") -> tuple[int, str, str]:
        """
        Send a JSON request to OAI NRF using HTTP/2 prior knowledge.

        The OAI NRF h2c endpoint closes plain HTTP/1.1 requests without a
        response. Python's standard clients do not support h2c prior knowledge,
        so use curl in the same way as the DCCF AMF/SMF subscription client.
        """
        cmd = [
            "curl",
            "--http2-prior-knowledge",
            "-sS",
            "-w",
            "\n%{http_code}",
            "-X",
            method,
            "-H",
            f"Content-Type: {content_type}",
            "-H",
            "Accept: application/json",
        ]

        if payload is not None:
            cmd.extend(["-d", json.dumps(payload)])

        cmd.append(url)

        result = subprocess.run(
            cmd,
            capture_output=True,
            text=True,
            timeout=10,
        )

        stdout = result.stdout or ""
        if "\n" in stdout:
            body, status_text = stdout.rsplit("\n", 1)
        else:
            body, status_text = "", stdout

        try:
            status_code = int(status_text.strip())
        except ValueError:
            status_code = 0

        if result.returncode != 0 and status_code == 0:
            raise RuntimeError(result.stderr.strip() or f"curl exited with {result.returncode}")

        return status_code, body, result.stderr

    def _build_nf_profile(self) -> dict:
        """Build 3GPP compliant NF Profile for registration"""
        return {
            "nfInstanceId": self.nf_instance_id,
            "nfType": self.nf_type,
            "nfStatus": "REGISTERED",
            "heartBeatTimer": self.heartbeat_interval,
            "ipv4Addresses": [self.ipv4_address] if self.ipv4_address else [],
            "nfServices": self._build_nf_services(),
            "nwdafInfo": {
                "eventIds": [
                    "NF_LOAD",
                    "UE_MOBILITY",
                    "UE_COMM",
                    "QOS_SUSTAINABILITY",
                    "ABNORMAL_BEHAVIOUR"
                ],
                "nwdafEvents": [
                    "NF_LOAD",
                    "UE_MOBILITY"
                ]
            }
        }

    def _build_nf_services(self) -> list:
        """Build NF Services list for registration"""
        services = []
        for svc in self.nf_services:
            service = {
                "serviceInstanceId": svc.get('serviceInstanceId', ''),
                "serviceName": svc.get('serviceName', ''),
                "versions": svc.get('versions', []),
                "scheme": svc.get('scheme', 'http'),
                "nfServiceStatus": "REGISTERED"
            }

            # Add IP endpoints
            ip_endpoints = svc.get('ipEndPoints', [])
            if ip_endpoints:
                normalized_endpoints = []
                for endpoint in ip_endpoints:
                    normalized_endpoint = dict(endpoint)
                    if not normalized_endpoint.get("ipv4Address") and self.ipv4_address:
                        normalized_endpoint["ipv4Address"] = self.ipv4_address
                    normalized_endpoints.append(normalized_endpoint)
                service["ipEndPoints"] = normalized_endpoints

            services.append(service)
        return services

    async def register(self) -> bool:
        """Register NWDAF-DCCF with NRF"""
        if not self.auto_register:
            logger.info("NRF auto-registration disabled")
            return False

        if not self.nrf_endpoint:
            logger.warning("NRF endpoint not configured")
            return False

        nf_profile = self._build_nf_profile()
        url = f"{self.nrf_endpoint}/nnrf-nfm/v1/nf-instances/{self.nf_instance_id}"

        logger.info(f"Registering with NRF at {url}")
        logger.debug(f"NF Profile: {nf_profile}")

        try:
            status_code, body, stderr = await asyncio.to_thread(
                self._curl_json,
                "PUT",
                url,
                nf_profile,
            )

            if status_code in [200, 201]:
                self._registered = True
                logger.info(f"Successfully registered with NRF (status: {status_code})")

                # Start heartbeat
                await self._start_heartbeat()
                return True

            logger.error(f"NRF registration failed: {status_code} - {body or stderr}")
            return False

        except Exception as e:
            logger.error(f"NRF registration error: {e}")
            return False

    async def deregister(self) -> bool:
        """Deregister from NRF"""
        if not self._registered:
            return True

        # Stop heartbeat
        if self._heartbeat_task:
            self._heartbeat_task.cancel()

        url = f"{self.nrf_endpoint}/nnrf-nfm/v1/nf-instances/{self.nf_instance_id}"

        try:
            status_code, body, stderr = await asyncio.to_thread(
                self._curl_json,
                "DELETE",
                url,
            )

            if status_code in [200, 204]:
                self._registered = False
                logger.info("Successfully deregistered from NRF")
                return True

            logger.error(f"NRF deregistration failed: {status_code} - {body or stderr}")
            return False

        except Exception as e:
            logger.error(f"NRF deregistration error: {e}")
            return False

    async def _start_heartbeat(self):
        """Start heartbeat task"""
        if self._heartbeat_task:
            self._heartbeat_task.cancel()
        self._heartbeat_task = asyncio.create_task(self._heartbeat_loop())

    async def _heartbeat_loop(self):
        """Send periodic heartbeats to NRF"""
        while self._registered:
            await asyncio.sleep(self.heartbeat_interval)

            if not self._registered:
                break

            url = f"{self.nrf_endpoint}/nnrf-nfm/v1/nf-instances/{self.nf_instance_id}"

            try:
                # PATCH with heartbeat update
                patch_data = [
                    {
                        "op": "replace",
                        "path": "/nfStatus",
                        "value": "REGISTERED"
                    }
                ]
                status_code, body, stderr = await asyncio.to_thread(
                    self._curl_json,
                    "PATCH",
                    url,
                    patch_data,
                    "application/json-patch+json",
                )

                if status_code in [200, 204]:
                    logger.debug("NRF heartbeat successful")
                elif status_code == 404:
                    logger.warning("NF not found in NRF, re-registering...")
                    await self.register()
                else:
                    logger.warning(f"NRF heartbeat failed: {status_code} - {body or stderr}")

            except Exception as e:
                logger.error(f"NRF heartbeat error: {e}")

    @property
    def is_registered(self) -> bool:
        return self._registered
